From 3cff302b2b11a925b04e3ff6f7b91ffcd45feb55 Mon Sep 17 00:00:00 2001
From: Adithya Srinivasan <adithyas@chromium.org>
Date: Fri, 10 Jan 2025 12:01:22 -0800
Subject: [PATCH] Reland "SpeculationRules: Use Traversal<HTMLScriptElement> instead of document.scripts()"

This is a reland of commit 14d9421e9f5f8b952934094f6afe1d4f097f34f4.
Minor change from the original patchset; uses the Document instead of
document_element_.

Original change's description:
> SpeculationRules: Use Traversal<HTMLScriptElement> instead of document.scripts()
>
> It's possible to run Document::UpdateBaseURL() while processing the
> removal of elements (ContainerNode::RemovedFrom), but before the
> HTMLCollections (like document.scripts()) are invalidated (which
> happens in ContainerNode::ChildrenChanged after all removals are
> complete).
>
> Bug: 386547460
> Change-Id: Ifc8036887dad3b999e77b1a014b950b97a20484a
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6148370
> Reviewed-by: Jeremy Roman <jbroman@chromium.org>
> Reviewed-by: Mason Freed <masonf@chromium.org>
> Commit-Queue: Adithya Srinivasan <adithyas@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1403679}

Bug: 386547460
Change-Id: I5bec86bc21e288878bd33120fb8601607a852a5b
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6165809
Commit-Queue: Adithya Srinivasan <adithyas@chromium.org>
Reviewed-by: Jeremy Roman <jbroman@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1404938}
---

--- a/third_party/blink/renderer/core/dom/document.cc
+++ b/third_party/blink/renderer/core/dom/document.cc
@@ -4587,9 +4587,9 @@
       anchor.InvalidateCachedVisitedLinkHash();
   }
 
-  for (Element* element : *scripts()) {
-    auto* script = To<HTMLScriptElement>(element);
-    script->Loader()->DocumentBaseURLChanged();
+  for (HTMLScriptElement& script :
+       Traversal<HTMLScriptElement>::DescendantsOf(*this)) {
+    script.Loader()->DocumentBaseURLChanged();
   }
 
   if (auto* document_rules = DocumentSpeculationRules::FromIfExists(*this)) {
--- a/third_party/blink/renderer/core/speculation_rules/speculation_rule_set_test.cc
+++ b/third_party/blink/renderer/core/speculation_rules/speculation_rule_set_test.cc
@@ -27,6 +27,7 @@
 #include "third_party/blink/renderer/core/html/html_anchor_element.h"
 #include "third_party/blink/renderer/core/html/html_area_element.h"
 #include "third_party/blink/renderer/core/html/html_base_element.h"
+#include "third_party/blink/renderer/core/html/html_collection.h"
 #include "third_party/blink/renderer/core/html/html_div_element.h"
 #include "third_party/blink/renderer/core/html/html_head_element.h"
 #include "third_party/blink/renderer/core/html/html_meta_element.h"
@@ -4413,5 +4414,30 @@
                           MatchesListOfURLs("https://example.com/bar")));
 }
 
+// Regression test for crbug.com/386547460.
+// This test will hit a DCHECK if we use document.scripts() when iterating
+// through scripts in |Document::UpdateBaseURL()| as the HTMLCollection's cached
+// node and count will not have been invalidated at that point, despite a
+// <script> having been removed.
+TEST_F(SpeculationRuleSetTest, Crbug386547460) {
+  base::HistogramTester histogram_tester;
+  DummyPageHolder page_holder;
+  Document& document = page_holder.GetDocument();
+  page_holder.GetFrame().GetSettings()->SetScriptEnabled(true);
+
+  document.documentElement()->setInnerHTML(R"html(
+    <head>
+      <script></script>
+    </head>
+    <body>
+      <script></script>
+      <base href="base.com"></base>
+    </body>
+  )html");
+
+  document.scripts()->item(2);
+  document.body()->setOuterHTML("<body>42</body>");
+}
+
 }  // namespace
 }  // namespace blink
